11/data/mail address

catenglaender opened the pull request:
This is an EmailAddress Data Type for more consistent and accurate validation.

Since ILIAS is used internationally, we are allowing many more characters than most Email Validators do. Umlauts ä, ö, ü and accented vowels become more and more common since updated standards technically allow them. Sadly, most existing checkers only let a subset of ASCII characters through. For this strict validation you can use isAscii() to verify that the email address is safe by old standards.

If isAscii() returns false, but the object did not throw an error on construction, it's a valid Unicode email address. However, most email servers (and php itself) cannot handle those optimally by default. One day when this Data Object is actually used for Email Addresses, PHP itself and the web server still need to be configured to handle Unicode strings correctly.

Commit:
11/EmailAddress Data Object

mjansenDatabay commented on components/ILIAS/Data/src/EmailAddress.php:
```diff
@@ -0,0 +1,223 @@
+    {
+        [, $domain] = explode('@', $address, 2);
+
+        $this->is_domainpart_ascii = $this->checkAscii($address);
```
Do we want to check the full `$address` or only the extracted `$domain` part here?

catenglaender replied:
Oh... You are absolutely right, it should be only the domain part. I will change that.

Right now, these checks have no real effect on anything, but maybe (in the future) admins could somewhere tell the instance what the email server can handle and require ASCII in only domain part, only in local parts or require ASCII in both, so the check has to be targeted.

catenglaender commented:
:heavy_check_mark: Rebased
:heavy_check_mark: Corrected domain part ASCII check
:heavy_check_mark: Corrected PHPStan check error in factory
Remaining failing unit tests are deprecation warnings not related to this PR

klees commented on components/ILIAS/Data/tests/EmailAddressTest.php:
```diff
@@ -0,0 +1,242 @@
+ */
+class EmailAddressTest extends TestCase
+{
+    #[DataProvider('validEmailProvider')] public function testEmailObjectConstruction(
```
Please move the `DataProvider` attributes to their own lines.

mjansenDatabay replied:
I agree, the PHP attribute should be moved to it's own line.

klees reviewed (approved):
Hi @catenglaender,

only a minor and maybe personal formatting request. The rest LGTM.

Kind regards!

klees commented:
Hi @mjansenDatabay,

would you be so kind and take care of a final review and approval, to follow the "not the same SP" rule? What's your stance on the formatting issue?

Thanks a lot!

mjansenDatabay commented:
Hi @klees / @catenglaender ,

yes, I will merge this after the formatting issue (and the merge conflict) has (have) been solved.

Best regards
Michael

catenglaender commented:
Thank you everyone! :slightly_smiling_face: 

formatting fixed
merge conflict solved

mjansenDatabay commented:
Thx @catenglaender
